tests/core/test_validation.py

```diff
@@ -955,3 +958,80 @@ def test_validate_axes_orthogonal(bias_index):
     assert np.array_equal(axes, axes_left)
     with pytest.raises(ValueError, match=msg):
         validate_axes(axes_left, must_be_orthogonal=True)
+
+
+@pytest.mark.parametrize('as_any', [True, False])
+@pytest.mark.parametrize('copy', [True, False])
+@pytest.mark.parametrize('dtype', [None, float])
+def test_cast_to_numpy(as_any, copy, dtype):
+    array_in = pyvista_ndarray([1, 2])
+    array_out = _cast_to_numpy(array_in, copy=copy, as_any=as_any, dtype=dtype)
+    assert np.array_equal(array_out, array_in)
+    if as_any:
+        assert type(array_out) is pyvista_ndarray
+    else:
+        assert type(array_out) is np.ndarray
+
+    if copy:
+        assert array_out is not array_in
+
+    if dtype is None:
+        assert array_out.dtype.type is array_in.dtype.type
+    else:
+        assert array_out.dtype.type is np.dtype(dtype).type
+
+
+def test_cast_to_numpy_raises():
+    msg = "Input cannot be cast as <class 'numpy.ndarray'>."
+    with pytest.raises(ValueError, match=msg):
+        _cast_to_numpy([[1], [2, 3]])
+
+
+def test_cast_to_numpy_must_be_real():
+    _ = _cast_to_numpy([0, 1], must_be_real=True)
+    _ = _cast_to_numpy("abc", must_be_real=False)
+
+    msg = "Array must have real numbers. Got dtype <class 'numpy.complex128'>"
+    with pytest.raises(TypeError, match=msg):
```

Comment:
Should dtype errors raise `ValueError` or `TypeError` ?

Reply:
Didn't think about that. Either seems justifiable. I guess I have a slight preference for TypeError.

Reply:
Type of object vs type of the object's values... not always clear. But yea here I think TypeError is appropriate for wrong dtype.